clusterloader2/pkg/test/simple_test_executor.go

```diff
@@ -126,105 +126,102 @@ func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) *util
+			ctx.GetState().Set(nsName, id, instances)
+			instancesStates = append(instancesStates, instances)
+		}
 
-			// Calculating maximal replica count of objects from object bundle.
-			var maxCurrentReplicaCount int32
-			for j := range instancesStates {
-				if instancesStates[j].CurrentReplicaCount > maxCurrentReplicaCount {
-					maxCurrentReplicaCount = instancesStates[j].CurrentReplicaCount
-				}
+		// Calculating maximal replica count of objects from object bundle.
+		var maxCurrentReplicaCount int32
+		for j := range instancesStates {
+			if instancesStates[j].CurrentReplicaCount > maxCurrentReplicaCount {
+				maxCurrentReplicaCount = instancesStates[j].CurrentReplicaCount
 			}
+		}
 
-			var namespaceWG wait.Group
-			// Deleting objects with index greater or equal requested replicas per namespace number.
-			// Objects will be deleted in reversed order.
-			for replicaCounter := phase.ReplicasPerNamespace; replicaCounter < maxCurrentReplicaCount; replicaCounter++ {
-				replicaIndex := replicaCounter
-				namespaceWG.Start(func() {
-					<-ticker.C
-					for j := len(phase.ObjectBundle) - 1; j >= 0; j-- {
-						if replicaIndex < instancesStates[j].CurrentReplicaCount {
-							if objectErrList := ste.ExecuteObject(ctx, &phase.ObjectBundle[j], nsName, replicaIndex, DELETE_OBJECT); !objectErrList.IsEmpty() {
-								errList.Concat(objectErrList)
-							}
+		// Deleting objects with index greater or equal requested replicas per namespace number.
+		// Objects will be deleted in reversed order.
+		for replicaCounter := phase.ReplicasPerNamespace; replicaCounter < maxCurrentReplicaCount; replicaCounter++ {
+			replicaIndex := replicaCounter
+			actions = append(actions, func() {
+				for j := len(phase.ObjectBundle) - 1; j >= 0; j-- {
+					if replicaIndex < instancesStates[j].CurrentReplicaCount {
+						if objectErrList := ste.ExecuteObject(ctx, &phase.ObjectBundle[j], nsName, replicaIndex, DELETE_OBJECT); !objectErrList.IsEmpty() {
+							errList.Concat(objectErrList)
 						}
 					}
-				})
-			}
-
-			// Calculating minimal replica count of objects from object bundle.
-			// If there is update operation to be executed, minimal replica count is set to zero.
-			minCurrentReplicaCount := int32(math.MaxInt32)
-			for j := range instancesStates {
-				if instancesStates[j].CurrentReplicaCount == phase.ReplicasPerNamespace {
-					minCurrentReplicaCount = 0
-					break
-				}
-				if instancesStates[j].CurrentReplicaCount < minCurrentReplicaCount {
-					minCurrentReplicaCount = instancesStates[j].CurrentReplicaCount
 				}
+			})
+		}
+
+		// Calculating minimal replica count of objects from object bundle.
+		// If there is update operation to be executed, minimal replica count is set to zero.
+		minCurrentReplicaCount := int32(math.MaxInt32)
+		for j := range instancesStates {
+			if instancesStates[j].CurrentReplicaCount == phase.ReplicasPerNamespace {
+				minCurrentReplicaCount = 0
```

Comment:
This makes me think that the logic here is not what we want.
I think that we should assume that "bundle is bundle".

If say bundle consists of objects A, B and C, there should be the same amount of those before and after the phase.
Otherwise, what does it mean that:
- you want to get to 5 replicas of bundle
- but there are 3 As, 5 Bs, and 7 Cs

For me - that's a bad input.

You're trying to solve it hackily by computing min and max, and iterating from desired to them, but that will cause more confusion to users than saying "bad input".

We should change that so that : if number of instances of particular objects within a bundle are different, then the input is bad, we return error and skip the phase.

Reply:
Probably should be done in a separate PR though

Reply:
This is a very strong assumption. Definitely this should be also mention in docs.

> Probably should be done in a separate PR though

I agree. I will create one after this one is merged.

Reply:
Yes - but if I'm saying that "this group of objects is a bundle", I don't want to assume that for some bundles there exist some subset of it, for some other superset and so on - this doesn't make any sense.

Filed https://github.com/kubernetes/perf-tests/issues/261 for it.